src/main/java/eu/dissco/orchestration/backend/security/JwtAuthConverter.java

```diff
@@ -2,33 +2,39 @@
 
 import jakarta.validation.constraints.NotNull;
 import java.util.Collection;
-import java.util.Optional;
-import java.util.stream.Collectors;
-import java.util.stream.Stream;
+import java.util.HashSet;
+import java.util.Map;
+import java.util.Set;
 import org.springframework.core.convert.converter.Converter;
 import org.springframework.security.authentication.AbstractAuthenticationToken;
 import org.springframework.security.core.GrantedAuthority;
 import org.springframework.security.oauth2.jwt.Jwt;
 import org.springframework.security.oauth2.jwt.JwtClaimNames;
 import org.springframework.security.oauth2.server.resource.authentication.JwtAuthenticationToken;
-import org.springframework.security.oauth2.server.resource.authentication.JwtGrantedAuthoritiesConverter;
 import org.springframework.stereotype.Component;
 
 @Component
 public class JwtAuthConverter implements Converter<Jwt, AbstractAuthenticationToken> {
 
-  private final JwtGrantedAuthoritiesConverter jwtGrantedAuthoritiesConverter = new JwtGrantedAuthoritiesConverter();
-
   @Override
   public AbstractAuthenticationToken convert(@NotNull Jwt jwt) {
-    Collection<GrantedAuthority> authorities =
-        converterToStream(jwt).collect(Collectors.toSet());
-    return new JwtAuthenticationToken(jwt, authorities, getPrincipalClaimName(jwt));
+    return new JwtAuthenticationToken(jwt, extractRoles(jwt), getPrincipalClaimName(jwt));
   }
 
-  private Stream<GrantedAuthority> converterToStream(Jwt jwt) {
-    return Optional.of(jwtGrantedAuthoritiesConverter.convert(jwt)).stream()
-        .flatMap(Collection::stream);
+  private Set<GrantedAuthority> extractRoles(Jwt jwt) {
+    Set<GrantedAuthority> authorities = new HashSet<>();
+    if (jwt.getClaims().containsKey("resource_access")) {
+      ((Map<String, Object>) jwt.getClaims().get("resource_access")).forEach((k, v) -> {
+        Map<String, Object> resourceAccess = (Map<String, Object>) v;
+        resourceAccess.forEach((k1, v1) -> {
+          if (k1.equals("roles")) {
+            ((Collection<String>) v1).forEach(
+                role -> authorities.add((GrantedAuthority) () -> "ROLE_" + role));
+          }
+        });
+      });
+    }
```

Comment:
Not sure how we're assigning roles, but would it be noteworthy to log if a user has no roles? Basically, will a regular user have a "regular user" role, or will they have no roles?

Reply:
They might have no roles and I think that is ok. So they can be authenticated and have no additional roles.